### PR TITLE
Implementing view model for rocket list fragment step

### PR DESCRIPTION
**Pull request by koba-wild**
Good day everyone. I have created fragments instead of activities and implemented ViewModels for fragments. Also, I would like to mention that I structured my architecture due to MVVM.

### CODE REVIEW

File: app/src/main/java/com/example/dragonx/view/fragments/RocketListFragment.kt [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+            recyclerView?.layoutManager = LinearLayoutManager(context)
+            val topSpacingItemDecoration = TopSpacingItemDecoration(30)
+            recyclerView?.addItemDecoration(topSpacingItemDecoration)
+            recyclerView?.adapter = RocketRecyclerAdapter(it, this)
+        })
+        return view
+    }
+
+    override fun onRocketClick(rocket: Rocket, position : Int) {
+        bundle.putInt("Rocket's number", position)
```
Comment (EvgenePrytula) [inline app/src/main/java/com/example/dragonx/view/fragments/RocketListFragment.kt:49]: What happens if bundle tag is not the same?

File: app/src/main/java/com/example/dragonx/viewmodel/RocketDetailsViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+import com.example.dragonx.models.Rocket
+
+import java.net.URL
+
+class RocketDetailsViewModel : ViewModel() {
+    var imageList = ArrayList<SlideModel>()
+
+
+    init {
+        Log.i("JsonParser", "JsonParser for RocketDetailsFragment was created")
```
Comment (EvgenePrytula) [inline app/src/main/java/com/example/dragonx/viewmodel/RocketDetailsViewModel.kt:19]: This tag from here?

File: app/src/main/java/com/example/dragonx/viewmodel/RocketDetailsViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+        rocket.year = jsonArray.string("first_flight")[rocketNumber]
+        return rocket
+    }
+
+    fun imagesForSlider(rocketNumber:Int?): ArrayList<SlideModel> {
+        when(rocketNumber) {
+            0 -> {  imageList.add(
+                SlideModel(
+                    "https://live.staticflickr.com/8578/16655995541_7817565ea9_k.jpg",
```
Comment (EvgenePrytula) [inline app/src/main/java/com/example/dragonx/viewmodel/RocketDetailsViewModel.kt:48]: Why do you use hardcoded urls? You have to parse it form json.

File: app/src/main/java/com/example/dragonx/viewmodel/RocketDetailsViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        rocket.description = jsonArray.string("description")[rocketNumber]
+        rocket.link = jsonArray.string("wikipedia")[rocketNumber]
+        rocket.height =
+            ("Diameter : meters : ${jsonArray.obj("diameter")[rocketNumber]?.double("meters")}," +
+                    " feet : ${jsonArray.obj("diameter")[rocketNumber]?.int("feet").toString()}")
+        rocket.mass =
+            ("Dry mass : ${jsonArray.int("dry_mass_kg")[rocketNumber]} (${jsonArray.int("dry_mass_lb")[0]}) lb")
+
+        rocket.year = jsonArray.string("first_flight")[rocketNumber]
+        return rocket
```
Comment (EvgenePrytula) [inline app/src/main/java/com/example/dragonx/viewmodel/RocketDetailsViewModel.kt:41]: Do you follow SOLID principles for code above?

File: app/src/main/java/com/example/dragonx/viewmodel/RocketDetailsViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    suspend fun parseJson (rocketNumber:Int?): Rocket {
+        lateinit var rocket: Rocket
+        lateinit var jsonArray: JsonArray<JsonObject>
+
+        val result = URL("https://api.spacexdata.com/v4/dragons").readText()
+        val parser: Parser = Parser()
+        val stringBuilder: StringBuilder = StringBuilder(result)
+        jsonArray = parser.parse(stringBuilder) as JsonArray<JsonObject>
```
Comment (EvgenePrytula) [inline app/src/main/java/com/example/dragonx/viewmodel/RocketDetailsViewModel.kt:29]: What happen it one of properties name will be changed? How do you handle errors?

File: app/src/main/java/com/example/dragonx/viewmodel/RocketListViewModel.kt [hunk 6]
```diff
@@ -1,23 +1,32 @@
...
-import java.net.URL
+import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import java.net.URL
 
-class JsonParser {
+class RocketListViewModel : ViewModel() {
     var rockets = arrayListOf<Rocket>()
     val images = arrayListOf("https://live.staticflickr.com/8578/16655995541_7817565ea9_k.jpg", "https://farm8.staticflickr.com/7647/16581815487_6d56cb32e1_b.jpg")
```
Comment (EvgenePrytula) [inline app/src/main/java/com/example/dragonx/viewmodel/RocketListViewModel.kt:19]: Hardcoded imeges?

File: app/src/main/java/com/example/dragonx/viewmodel/RocketListViewModel.kt [hunk 7]
```diff
@@ -1,23 +1,32 @@
...
-    private lateinit var jsonArray: JsonArray<JsonObject>
+
+    init {
+        Log.i("JsonParser", "JsonParser for RocketListFragment was created")
+    }
 
     private fun parseJson () : JsonArray<JsonObject> {
         val result = URL("https://api.spacexdata.com/v4/dragons").readText()
         val parser: Parser = Parser()
         val stringBuilder: StringBuilder = StringBuilder(result)
```
Comment (EvgenePrytula) [inline app/src/main/java/com/example/dragonx/viewmodel/RocketListViewModel.kt:28]: Explain why did you use StringBuilder instead of StringBuffer?

**Review by EvgenePrytula — CHANGES_REQUESTED** [review]
Need to be improved. Project structure is wrong and doesn't accord clean code principles. Absence of using Solid principles.

File: app/src/main/AndroidManifest.xml [hunk 8]
```diff
@@ -14,22 +14,22 @@
         android:theme="@style/Theme.DragonX">
         <activity android:name=".RocketDetails"
             android:label="@string/title_activity_rocket_details"
-            android:parentActivityName=".MainActivity">
+            android:parentActivityName=".view.activities.MainActivity">
             <meta-data
                 android:name="android.support.PARENT_ACTIVITY"
```
[the remark this answers is not in the document]
Comment (koba-wild) [reply app/src/main/AndroidManifest.xml:19]: Cleared the Android manifest file

File: app/src/main/java/com/example/dragonx/view/activities/MainActivity.kt [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+import com.example.dragonx.R
+import com.example.dragonx.view.fragments.RocketDetailsFragment
+import com.example.dragonx.view.fragments.RocketListFragment
+
+class MainActivity : AppCompatActivity() {
+    private val fragmentRocketList = RocketListFragment()
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+        supportFragmentManager.beginTransaction().add(R.id.fragment_rocket_list,
```
[the remark this answers is not in the document]
Comment (koba-wild) [reply app/src/main/java/com/example/dragonx/view/activities/MainActivity.kt:15]: Replaced it with replace)

File: app/src/main/java/com/example/dragonx/view/fragments/RocketDetailsFragment.kt [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        val view = inflater.inflate(R.layout.fragment_rocket_details, container, false)
+        var rocketNumber = arguments?.getInt("Rocket's number")
+        viewModel = ViewModelProvider(this).get(RocketDetailsViewModel::class.java)
+
+        myScope.launch {
+            rocket.postValue(viewModel.parseJson(rocketNumber))
+        }
+
+        rocket.observe(viewLifecycleOwner, {
+            view.rocketName.text = it.name
```
[the remark this answers is not in the document]
Comment (koba-wild) [reply app/src/main/java/com/example/dragonx/view/fragments/RocketDetailsFragment.kt:38]: Done. used https://developer.android.com/topic/libraries/view-binding#fragments materials

File: app/src/main/java/com/example/dragonx/view/fragments/RocketDetailsFragment.kt [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+import com.denzcoskun.imageslider.models.SlideModel
+import com.example.dragonx.R
+import com.example.dragonx.models.Rocket
+import com.example.dragonx.viewmodel.RocketDetailsViewModel
+import com.example.dragonx.viewmodel.RocketListViewModel
+import kotlinx.android.synthetic.main.fragment_rocket_details.view.*
+import kotlinx.coroutines.*
+
+class RocketDetailsFragment : Fragment() {
+    private lateinit var viewModel: RocketDetailsViewModel
```
Comment (svitlanakrav) [inline app/src/main/java/com/example/dragonx/view/fragments/RocketDetailsFragment.kt:20]: why you don't initial viewmodel here?

File: app/src/main/java/com/example/dragonx/view/fragments/RocketDetailsFragment.kt [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+    private val rocket = MutableLiveData<Rocket>()
+    private val myJob = Job()
+    private val myScope = CoroutineScope(Dispatchers.IO + myJob)
+
+    override fun onCreateView(
+        inflater: LayoutInflater, container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View? {
+        val view = inflater.inflate(R.layout.fragment_rocket_details, container, false)
+        var rocketNumber = arguments?.getInt("Rocket's number")
```
Comment (svitlanakrav) [inline app/src/main/java/com/example/dragonx/view/fragments/RocketDetailsFragment.kt:30]: use const value for naming of bundle data

File: app/src/main/java/com/example/dragonx/view/fragments/RocketListFragment.kt [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+import androidx.recyclerview.widget.LinearLayoutManager
+import androidx.recyclerview.widget.RecyclerView
+import com.example.dragonx.*
+import com.example.dragonx.models.Rocket
+import com.example.dragonx.viewmodel.RocketListViewModel
+import kotlinx.coroutines.*
+
+
+class RocketListFragment : Fragment(), OnRocketClickListener {
+    private val bundle = Bundle()
```
Comment (svitlanakrav) [inline app/src/main/java/com/example/dragonx/view/fragments/RocketListFragment.kt:21]: I saw about that, don't use global variable if it is not justified.

File: app/src/main/java/com/example/dragonx/view/fragments/RocketDetailsFragment.kt (hunk 11 above)
Comment (koba-wild) [reply app/src/main/java/com/example/dragonx/view/fragments/RocketDetailsFragment.kt:20]: I just stuck the tutorial video on Udacity )

File: app/src/main/java/com/example/dragonx/viewmodel/RocketDetailsViewModel.kt [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+        rocket = Rocket()
+        rocket.name = jsonArray.string("name")[rocketNumber!!]
```
Comment (svitlanakrav) [inline app/src/main/java/com/example/dragonx/viewmodel/RocketDetailsViewModel.kt:31]: try to not use !!. Better to use ? or let

File: app/src/main/java/com/example/dragonx/view/fragments/RocketDetailsFragment.kt (hunk 12 above)
Comment (koba-wild) [reply app/src/main/java/com/example/dragonx/view/fragments/RocketDetailsFragment.kt:30]: Will do

File: app/src/main/java/com/example/dragonx/viewmodel/RocketDetailsViewModel.kt [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+        rocket.description = jsonArray.string("description")[rocketNumber]
+        rocket.link = jsonArray.string("wikipedia")[rocketNumber]
+        rocket.height =
+            ("Diameter : meters : ${jsonArray.obj("diameter")[rocketNumber]?.double("meters")}," +
```
Comment (svitlanakrav) [inline app/src/main/java/com/example/dragonx/viewmodel/RocketDetailsViewModel.kt:35]: What will you do if you will need to use height with another format? For example : " Diameter ${diameter}"

File: app/src/main/java/com/example/dragonx/viewmodel/RocketDetailsViewModel.kt [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+                    " feet : ${jsonArray.obj("diameter")[rocketNumber]?.int("feet").toString()}")
+        rocket.mass =
+            ("Dry mass : ${jsonArray.int("dry_mass_kg")[rocketNumber]} (${jsonArray.int("dry_mass_lb")[0]}) lb")
+
+        rocket.year = jsonArray.string("first_flight")[rocketNumber]
+        return rocket
+    }
+
+    fun imagesForSlider(rocketNumber:Int?): ArrayList<SlideModel> {
```
Comment (svitlanakrav) [inline app/src/main/java/com/example/dragonx/viewmodel/RocketDetailsViewModel.kt:44]: Why do you return list if this list don't create in this function?

**Review by svitlanakrav — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/dragonx/view/fragments/RocketDetailsFragment.kt (hunk 11 above)
Comment (svitlanakrav) [reply app/src/main/java/com/example/dragonx/view/fragments/RocketDetailsFragment.kt:20]: In this situation you reinitial viewmodel when view recreate( f.e. when app rotate). Are you sure you need there?

File: app/src/main/java/com/example/dragonx/presentation/RocketDetails/RocketDetailsFragment.kt [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+import com.example.dragonx.R
+import kotlinx.android.synthetic.main.fragment_rocket_details.view.*
+import com.example.dragonx.models.Rocket
+import com.example.dragonx.util.Constants.Companion.ROCKET_NUMBER
+import com.example.dragonx.viewmodel.ViewModelFactory
+
+class RocketDetailsFragment : Fragment() {
+
+    companion object {
+        private const val TAG = "KobaLOG"
```
Comment (svitlanakrav) [inline app/src/main/java/com/example/dragonx/presentation/RocketDetails/RocketDetailsFragment.kt:20]: It is better to write the place where the logs come from. Like RocketFragment

File: app/src/main/java/com/example/dragonx/presentation/RocketList/RocketListFragment.kt [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+        val recyclerView = view.findViewById<RecyclerView>(R.id.recycler_view)
+        val errorPicture = view.findViewById<ImageView>(R.id.imageForError)
+
+        recyclerView.layoutManager = LinearLayoutManager(context)
+        val topSpacingItemDecoration = TopSpacingItemDecoration(30)
+        recyclerView.addItemDecoration(topSpacingItemDecoration)
+        val viewModel = ViewModelProvider(this).get(RocketListViewModel::class.java)
+
+        viewModel.rocketsData.observe(viewLifecycleOwner, {
+            recyclerView.adapter = RocketRecyclerAdapter(it, this)
```
Comment (svitlanakrav) [inline app/src/main/java/com/example/dragonx/presentation/RocketList/RocketListFragment.kt:47]: You create new adapter every time when data observe

**Pull request closed without merging**